### ISSUE TITLE
Add early stopping criteria

### ISSUE THREAD

**Issue by cgsavard** [hqucms/weaver#11]
Currently there is no early stopping criteria which can lead to issues with overtraining as seen here in a GNN that I am training with weaver.
<img width="367" alt="Screen Shot 2022-01-31 at 11 11 54 AM" src="https://user-images.githubusercontent.com/15660721/151775234-99676a5e-4b1a-4d57-87e9-3a35d3046ac2.png">

I have developed my own early stopping criteria in my forked branch [here](https://github.com/cgsavard/weaver/commit/c8c0640dd1ec135a394efc14a1927c93b435651c) which stops training after the validation loss doesn't decreases in n (user defined) epochs. I think that this would be very useful for other weaver users and would be happy to put a PR in for this feature. Let me know what you think.

**Comment by hqucms** [comment]
Hi @cgsavard -- I think if you see overfitting like this, a better solution would be to see if you can increase the training sample size, or reduce the model complexity, or add dropout / weight decay, or simply reduce the number of epochs, etc. Early stopping is not really solving the problem -- it just bypasses it. And it will not work well with any learning rate schedulers.
